### ISSUE TITLE
"node-gyp rebuild"

### ISSUE THREAD

**Issue by 1NAX2** [nteract/hydrogen#458]
npm ERR! zmq@2.15.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zmq@2.15.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zmq package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zmq
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zmq
npm ERR! There is likely additional logging output above.

**Comment by lgeiger** [comment]
Have you tried the solutions suggested in our [troubleshooting guide](https://github.com/nteract/hydrogen/blob/master/TROUBLESHOOTING.md)?

Side note: Any updates on #457?

**Comment by 1NAX2** [comment]
> Side note: Any updates on #457?

i no longer get that error

**Comment by 1NAX2** [comment]
npm ERR! zmq@2.15.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zmq@2.15.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zmq package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zmq
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zmq
npm ERR! There is likely additional logging output above.
I am still receiving this error
When doing apm --version i get
$ apm --version
apm  1.12.5
npm  3.10.5
node 4.4.5
python 2.7.2
git 1.9.4.msysgit.2
visual studio 2015
Can anyone help?
i have attempted to switch it from VS2015 to VS2013 but it still shows VS2015 i have rebooted my machine too.

**Comment by lgeiger** [comment]
Are you using the default windows command prompt?

One thing worth trying is `apm install hydrogen --msvs_version=2013`

**Comment by 1NAX2** [comment]
I am using git bash

**Comment by 1NAX2** [comment]
apm install hydrogen --msvs_version=2013 returns the same error

**Comment by lgeiger** [comment]
Try [this comment](https://github.com/nteract/hydrogen/issues/439#issuecomment-249676388).

**Comment by 1NAX2** [comment]
Installing hydrogen to C:\Users\Conor\.atom\packages failed

> zmq@2.15.3 install C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116828-6656-58su2k\node_modules\Hydrogen\node_modules\zmq
> node-gyp rebuild


C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116828-6656-58su2k\node_modules\Hydrogen\node_modules\zmq>if not defined npm_config_node_gyp (node "c:\Users\Conor\AppData\Local\atom\app-1.10.2\resources\app\apm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116828-6656-58su2k
`-- (empty)


module.js:327
    throw err;
    ^

Error: Cannot find module 'c:\Users\Conor\AppData\Local\atom\app-1.10.2\resources\app\apm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Function.Module.runMain (module.js:441:10)
    at startup (node.js:139:18)
    at node.js:968:3
npm WARN enoent ENOENT: no such file or directory, open 'C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116828-6656-58su2k\package.json'
npm WARN apm-install-dir-116828-6656-58su2k No description
npm WARN apm-install-dir-116828-6656-58su2k No repository field.
npm WARN apm-install-dir-116828-6656-58su2k No README data
npm WARN apm-install-dir-116828-6656-58su2k No license field.
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "c:\\Users\\Conor\\AppData\\Local\\atom\\app-1.10.2\\resources\\app\\apm\\bin\\node.exe" "c:\\Users\\Conor\\AppData\\Local\\atom\\app-1.10.2\\resources\\app\\apm\\node_modules\\npm\\bin\\npm-cli.js" "--globalconfig" "C:\\Users\\Conor\\.atom\\.apm\\.apmrc" "--userconfig" "C:\\Users\\Conor\\.atom\\.apmrc" "install" "C:\\Users\\Conor\\AppData\\Local\\Temp\\d-116828-6656-1k0m7hx\\package.tgz" "--runtime=electron" "--target=0.37.8" "--arch=ia32" "--global-style" "--msvs_version=2013"
npm ERR! node v4.4.5
npm ERR! npm  v3.10.5
npm ERR! code ELIFECYCLE

npm ERR! zmq@2.15.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zmq@2.15.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zmq package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zmq
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zmq
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116828-6656-58su2k\npm-debug.log
npm ERR! code 1
no luck yet again ;-;

**Comment by 1NAX2** [comment]
I have updated stuff in the hopes it works:
Current:
$ apm --version
apm  1.12.5
npm  3.10.5
node 4.4.5
python 2.7.2
git 2.10.0.windows.1
visual studio 2015

This is what the trouble shooting guide says i need:
apm  1.12.5
npm  3.10.5
node 4.4.5
python 2.7.12
git 2.8.1.windows.1
visual studio 2015

im really running out of ideas, ;-;

**Comment by n-riesco** [comment]
On 28/09/16 17:44, Naxane wrote:
> I am using git bash

Does the installation work when you run Atom from cmd.exe?

**Comment by 1NAX2** [comment]
installing hydrogen for atom through the apm command on cmd does not work
either, i get the same code, i doubt its a problem with gitbash then

**Comment by 1NAX2** [comment]
i attempted to install term3, and got the same error relating to node-gyp rebuild
npm ERR! ptyw.js@0.4.1 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the ptyw.js@0.4.1 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the ptyw.js package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs ptyw.js
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls ptyw.js
npm ERR! There is likely additional logging output above.

**Comment by n-riesco** [comment]
On 30/09/16 16:19, Naxane wrote:
> i attempted to install term3, and got the same error relating to node-gyp rebuild
> npm ERR! ptyw.js@0.4.1 install: |node-gyp rebuild|
> npm ERR! Exit status 1
> npm ERR!
> npm ERR! Failed at the ptyw.js@0.4.1 install script 'node-gyp rebuild'.
> npm ERR! Make sure you have the latest version of node.js and npm installed.
> npm ERR! If you do, this is most likely a problem with the ptyw.js package,
> npm ERR! not with npm itself.
> npm ERR! Tell the author that this fails on your system:
> npm ERR! node-gyp rebuild
> npm ERR! You can get information on how to open an issue for this project with:
> npm ERR! npm bugs ptyw.js
> npm ERR! Or if that isn't available, you can get their info via:
> npm ERR! npm owner ls ptyw.js
> npm ERR! There is likely additional logging output above.

This part of the log isn't very informative.

---

Please, correct me if I'm wrong. The current situation (as I gather messages in this issue) is:
- `node-gyp rebuild` fails 
- Python 2.7 is installed and accessible by Atom
- Visual Studio 2015 is installed and accessible by Atom
- Are there any other versions of Visual Studio also installed?
- One of the log messages complains about a missing file: `c:\Users\Conor\AppData\Local\atom\app-1.10.2\resources\app\apm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js`

These are the things I'd check:

1) Is the file `c:\Users\Conor\AppData\Local\atom\app-1.10.2\resources\app\apm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js` really missing?

2) In Visual Studio 2015, Visual C++ is not installed by default. From https://msdn.microsoft.com/en-gb/library/60k1461a.aspx :
```
In Visual Studio 2015, Visual C++ is not installed by default. When installing, be sure to choose Custom installation and then choose the C++ components you require. Or, if Visual Studio is already installed, choose File | New | Project | C++ and you will be prompted to install the necessary components.
```

3) If both, the `node-gyp.js` file and Visual C++ are present in yout machine, then I'd run this:
```
$ apm uninstall hydrogen
$ apm install hydrogen
```
and post all the output here.

**Comment by 1NAX2** [comment]
- `node-gyp rebuild` fails - True
- Python 2.7 is installed and accessible by Atom - True
- Visual Studio 2015 is installed and accessible by Atom - True
- Are there any other versions of Visual Studio also installed? - 2015 and 2013 versions are both installed, 2015 is seen as the one accessible by Atom
- One of the log messages complains about a missing file: `c:\Users\Conor\AppData\Local\atom\app-1.10.2\resources\app\apm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js`
![image](https://cloud.githubusercontent.com/assets/7032300/19003337/413f8240-8748-11e6-9d34-fc43487a4346.png)

**Comment by n-riesco** [comment]
@1NAX2 Could you check whether installing Visual C++ in Visual Studio 2015 fixes the problem?

From https://msdn.microsoft.com/en-gb/library/60k1461a.aspx :

> In Visual Studio 2015, Visual C++ is not installed by default. When installing, be sure to choose Custom installation and then choose the C++ components you require. Or, if Visual Studio is already installed, choose File | New | Project | C++ and you will be prompted to install the necessary components.

**Comment by 1NAX2** [comment]
ive installed C# and C++ in VS2015
uninstalled hydrogen, which returned an error because it doesnt exist, and the reinstalled it, which raised an error:
Conor@CONOR-PC /c/Sites
$ apm install hydrogen
Installing hydrogen to C:\Users\Conor\.atom\packages failed

> zmq@2.15.3 install C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-11920-vuakbm\node_modules\Hydrogen\node_modules\zmq
> node-gyp rebuild

C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-11920-vuakbm
`-- (empty)


gyp ERR! configure error
gyp ERR! stack Error: Python executable "c:\Users\Conor\AppData\Local\Programs\Python\Python35-32\python.EXE" is v3.5.2, which is not supported by gyp.
gyp ERR! stack You can pass the --python switch to point to Python >= v2.5.0 & < 3.0.0.
gyp ERR! stack     at failPythonVersion (c:\Users\Conor\AppData\Roaming\npm\node_modules\node-gyp\lib\configure.js:454:14)
gyp ERR! stack     at c:\Users\Conor\AppData\Roaming\npm\node_modules\node-gyp\lib\configure.js:443:9
gyp ERR! stack     at ChildProcess.exithandler (child_process.js:204:7)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at maybeClose (internal/child_process.js:829:16)
gyp ERR! stack     at Socket.<anonymous> (internal/child_process.js:319:11)
gyp ERR! stack     at emitOne (events.js:77:13)
gyp ERR! stack     at Socket.emit (events.js:169:7)
gyp ERR! stack     at Pipe._onclose (net.js:486:12)
gyp ERR! System Windows_NT 10.0.14393
gyp ERR! command "c:\\Program Files\\nodejs\\node.exe" "c:\\Users\\Conor\\AppData\\Roaming\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-11920-vuakbm\node_modules\Hydrogen\node_modules\zmq
gyp ERR! node -v v4.5.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
npm WARN enoent ENOENT: no such file or directory, open 'C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-11920-vuakbm\package.json'
npm WARN apm-install-dir-116830-11920-vuakbm No description
npm WARN apm-install-dir-116830-11920-vuakbm No repository field.
npm WARN apm-install-dir-116830-11920-vuakbm No README data
npm WARN apm-install-dir-116830-11920-vuakbm No license field.
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "c:\\Users\\Conor\\AppData\\Local\\atom\\app-1.10.2\\resources\\app\\apm\\bin\\node.exe" "c:\\Users\\Conor\\AppData\\Local\\atom\\app-1.10.2\\resources\\app\\apm\\node_modules\\npm\\bin\\npm-cli.js" "--globalconfig" "C:\\Users\\Conor\\.atom\\.apm\\.apmrc" "--userconfig" "C:\\Users\\Conor\\.atom\\.apmrc" "install" "C:\\Users\\Conor\\AppData\\Local\\Temp\\d-116830-11920-jdf52x\\package.tgz" "--runtime=electron" "--target=0.37.8" "--arch=ia32" "--global-style" "--msvs_version=2015"
npm ERR! node v4.4.5
npm ERR! npm  v3.10.5
npm ERR! code ELIFECYCLE

npm ERR! zmq@2.15.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zmq@2.15.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zmq package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zmq
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zmq
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-11920-vuakbm\npm-debug.log
npm ERR! code 1

**Comment by n-riesco** [comment]
@1NAX2 OK, there's some progress there. Now the problem is that `apm` is using Python 3 instead of Python 2. You need to find the location of the executable for Python 2 and run `apm config set python c:\path\to\python.exe`.

**Comment by 1NAX2** [comment]
i didn't used to get the python issue, but i've done that command anyway,

Conor@CONOR-PC /
$ apm install hydrogen
Installing hydrogen to C:\Users\Conor\.atom\packages failed

> zmq@2.15.3 install C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-10552-1inrv2i\node_modules\Hydrogen\node_modules\zmq
> node-gyp rebuild

Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
TRACKER : error TRK0005: Failed to locate: "CL.exe". The system cannot find the file specified. [C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-10552-1inrv2i\node_modules\Hydrogen\node_modules\zmq\build\zmq.vcxproj]


C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-10552-1inrv2i
`-- (empty)


gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (c:\Users\Conor\AppData\Roaming\npm\node_modules\node-gyp\lib\build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Windows_NT 10.0.14393
gyp ERR! command "c:\\Program Files\\nodejs\\node.exe" "c:\\Users\\Conor\\AppData\\Roaming\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-10552-1inrv2i\node_modules\Hydrogen\node_modules\zmq
gyp ERR! node -v v4.5.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
npm WARN enoent ENOENT: no such file or directory, open 'C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-10552-1inrv2i\package.json'
npm WARN apm-install-dir-116830-10552-1inrv2i No description
npm WARN apm-install-dir-116830-10552-1inrv2i No repository field.
npm WARN apm-install-dir-116830-10552-1inrv2i No README data
npm WARN apm-install-dir-116830-10552-1inrv2i No license field.
npm ERR! Windows_NT 10.0.14393
npm ERR! argv "c:\\Users\\Conor\\AppData\\Local\\atom\\app-1.10.2\\resources\\app\\apm\\bin\\node.exe" "c:\\Users\\Conor\\AppData\\Local\\atom\\app-1.10.2\\resources\\app\\apm\\node_modules\\npm\\bin\\npm-cli.js" "--globalconfig" "C:\\Users\\Conor\\.atom\\.apm\\.apmrc" "--userconfig" "C:\\Users\\Conor\\.atom\\.apmrc" "install" "C:\\Users\\Conor\\AppData\\Local\\Temp\\d-116830-10552-wpmah\\package.tgz" "--runtime=electron" "--target=0.37.8" "--arch=ia32" "--global-style" "--msvs_version=2015"
npm ERR! node v4.4.5
npm ERR! npm  v3.10.5
npm ERR! code ELIFECYCLE

npm ERR! zmq@2.15.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zmq@2.15.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zmq package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zmq
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zmq
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Conor\AppData\Local\Temp\apm-install-dir-116830-10552-1inrv2i\npm-debug.log
npm ERR! code 1

**Comment by n-riesco** [comment]
@1NAX2 The log complains that `CL.exe` can't be found. `CL.exe` is C++ compiler that you've just installed. Could you check whether `CL.exe` is accessible in the command line? What do you get when you run `cl` in the command line?

**Comment by 1NAX2** [comment]
there is no such thing as "cl", what is it and how can i properly get it?

**Comment by n-riesco** [comment]
Did you try this?

> Or, if Visual Studio is already installed, choose File | New | Project | C++ and you will be prompted to install the necessary components.

**Comment by n-riesco** [comment]
@1NAX2 [This](https://blogs.msdn.microsoft.com/deva/2015/03/24/after-migration-vs-2013-c-project-throws-error-trk0005-failed-to-locate-cl-exe-the-system-cannot-find-the-file-specified/) might be also relevant.

**Comment by 1NAX2** [comment]
![image](https://cloud.githubusercontent.com/assets/7032300/19006453/71327570-8757-11e6-8cbe-68a47a649288.png)
Are these the components i'm installing?

**Comment by n-riesco** [comment]
[Here](https://blogs.msdn.microsoft.com/vcblog/2015/07/24/setup-changes-in-visual-studio-2015-affecting-c-developers/) suggests to select the MFC template.

**Comment by 1NAX2** [comment]
Installing MFC fixed it

**Comment by n-riesco** [comment]
@1NAX2 Thanks a lot for reporting back! We should update `README.md` with the steps to install Visual C++.

**Comment by lgeiger** [comment]
Closed by #476
